core/src/cgcloud/core/box.py

```diff
@@ -319,18 +320,29 @@ def prepare( self, ec2_keypair_globs,
         if instance_type is None:
             instance_type = self.recommended_instance_type( )
 
-        virtualization_type = self.__get_virtualization_type( instance_type, virtualization_type )
+        virtualization_types = self.__get_virtualization_types( instance_type, virtualization_type )
 
+        image = None
         if image_ref is not None:
             image = self.__select_image( image_ref )
         else:
-            log.info( "Looking up default image for role %s, ... ", self.role( ) )
-            image = self._base_image( virtualization_type )
-            log.info( "... found %s.", image.id )
+            for type in virtualization_types:
+                log.info( "Looking up default image for role %s and type %s, ... ", self.role( ), type )
+                try:
+                    image = self._base_image( type )
+                except RuntimeError:
+                    log.warning("... did not find a suitable image.")
+                else:
+                    log.info( "... found %s.", image.id )
+                if image is not None and image.virtualization_type in virtualization_types:
+                    break
+
+        if image is None:
+            raise RuntimeError("Could not find suitable image for role %s", type)
 
-        if image.virtualization_type != virtualization_type:
+        if image.virtualization_type not in virtualization_types:
             raise RuntimeError( "Expected virtualization type %s but image only supports %s" % (
```

Comment:
Also:

```
>>> "%s" % ['hvm','paravirtual']
"['hvm', 'paravirtual']"
```

Can we make that a bit more like an English sentence. It would be overkill just for this but I need this all the time. Maybe we can be inspired by

https://pypi.python.org/pypi/inflect#converting-lists-of-words-to-phrases

It would be fun to write such a method and add it to bd2k-python-lib.